app.go

```diff
@@ -24,142 +24,238 @@ import (
+
+// Timeout is a convenience function to construct a context with a timeout.
+// It's only intended to reduce noise in the main function; since it doesn't
+// expose context.CancelFunc, it may leak resources.
+func Timeout(d time.Duration) context.Context {
+	ctx, cancel := context.WithTimeout(context.Background(), d)
+	// Assign to the blank identifier on a separate line to convince the linter
+	// that we really don't want the cancel function.
+	_ = cancel
+	return ctx
+}
+
+// An Option configures an App.
+type Option interface {
+	apply(*App)
+}
+
+type optionFunc func(*App)
+
+func (f optionFunc) apply(app *App) { f(app) }
+
+// Provide registers constructors with the application's dependency injection
+// container. Constructors provide one or more types, can depend on other
+// types available in the container, and may optionally return an error. For
+// example:
+//
+//  // Provides type *C, depends on *A and *B.
+//  func(*A, *B) *C
+//
+//  // Provides type *C, depends on *A and *B, and indicates failure by
+//  // returning an error.
+//  func(*A, *B) (*C, error)
+//
+//  // Provides type *B and *C, depends on *A, and can fail.
+//  func(*A) (*B, *C, error)
+//
+// The order in which constructors are provided doesn't matter. Constructors
+// are called lazily and their results are cached for reuse.
+//
+// Taken together, these properties make it perfectly reasonable to Provide a
+// large number of standard constructors even if only a fraction of them are
+// used.
+//
+// See the documentation for go.uber.org/dig for further details.
+func Provide(constructors ...interface{}) Option {
+	return optionFunc(func(app *App) {
+		for _, c := range constructors {
+			app.provide(c)
+		}
+	})
+}
+
+// Invoke registers functions that are executed eagerly on application start.
+// Arguments for these functions are provided from the application's
+// dependency injection container, so the arguments form the leaves of the
+// dependency resolution graph. That is, only constructors whose returned
+// types are required to invoke these functions are executed.
+//
+// Unlike constructors, invoked functions are always executed, and they're
+// always run in order. Invoked functions may have any number of returned
+// values. If the final returned object is an error, it's assumed to be a
+// success indicator. All other returned values are discarded.
+//
+// See the documentation for go.uber.org/dig for further details.
+func Invoke(funcs ...interface{}) Option {
+	return optionFunc(func(app *App) {
+		app.invokes = append(app.invokes, funcs...)
+	})
+}
+
+// Options composes a collection of Options into a single Option.
+func Options(opts ...Option) Option {
+	return optionFunc(func(app *App) {
+		for _, opt := range opts {
+			opt.apply(app)
+		}
+	})
+}
+
+// An App is a modular application built around dependency injection.
 type App struct {
+	optionErr error
 	container *dig.Container
 	lifecycle *lifecycle
+	invokes   []interface{}
```

Comment:
At one point we had `dig.Fn` as an alias for `interface{}`, to make room for better documentation and make it clearer for usages just like this.

Now looking at this I think `dig.Invocable` or something like that might also be on the cards.

Nothing to do with the PR, just thinking out loud.